File: Easy/108. Convert Sorted Array to Binary Search Tree/ans.js
var sortedArrayToBST = function(nums, start=0, end = nums.length-1) {
    if(start<=end){
        let mid = Math.floor((start+end)/2);
        // console.log(mid, end)
        let root = new TreeNode(nums[mid]);
        root.left = sortedArrayToBST(nums, start, mid-1);
        root.right = sortedArrayToBST(nums, mid+1, end);
        return root;
    }
    return null;
};

File: Easy/108. Convert Sorted Array to Binary Search Tree/ans.py
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if not nums: return None
        
        # Each item in the stack holds: floor, ceiling, parent and parent's direction.
        stack = [(0, len(nums) - 1, None, None)]
        root = None
        while stack:
            floor, ceiling, parent, direction = stack.pop()
            middleIndex = (floor + ceiling) // 2
            n = TreeNode(nums[middleIndex])
            
            # Add the root node. parent should only ever be null in this one case.
            if not parent:
                root = n
            else:
                if direction == 'l':
                    parent.left = n
                elif direction == 'r':
                    parent.right = n
            
            # Add the next items to the stack, as necessary. Add right first.
			# Similar to the recursive approach, both floor and ceiling should never cross.
            if middleIndex + 1 <= ceiling:
                stack.append((middleIndex + 1, ceiling, n, 'r'))
            if floor <= middleIndex - 1:
                stack.append((floor, middleIndex - 1, n, 'l'))
        return root
